toolkit/tools/imagegen/configuration/fileconfig.go

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+//
+
+package configuration
+
+import (
+	"encoding/json"
+	"fmt"
+)
+
+// DestinationFileConfigList is a list of destination files to copy.
+type FileConfigList []FileConfig
+
+// FileConfig specifies options for how a file is copied in the target OS.
+type FileConfig struct {
+	// The file path in the target OS that the file will be copied to.
+	Path string `json:"Path"`
+
+	// The file permissions to set on the file.
+	Permissions *FilePermissions `json:"Permissions"`
+}
+
+var (
+	DefaultFileConfig = FileConfig{
+		Path:        "",
+		Permissions: nil,
+	}
+)
+
+func (l *FileConfigList) IsValid() (err error) {
+	if len(*l) <= 0 {
+		return fmt.Errorf("list is empty")
+	}
+
+	return nil
+}
+
+func (l *FileConfigList) UnmarshalJSON(b []byte) error {
+	var err error
+
+	// Try to parse as a single value.
+	var fileConfig FileConfig
+	err = json.Unmarshal(b, &fileConfig)
+	if err == nil {
+		*l = FileConfigList{fileConfig}
+		return nil
+	}
+
+	// Try to parse as a list.
+	type IntermediateTypeFileConfigList FileConfigList
+	err = json.Unmarshal(b, (*IntermediateTypeFileConfigList)(l))
+	if err != nil {
+		return fmt.Errorf("failed to parse [FileConfigList]: %w", err)
+	}
+
+	// Validate the unmarshaled object.
+	err = l.IsValid()
+	if err != nil {
+		return fmt.Errorf("failed to parse [FileConfigList]: %w", err)
+	}
+
+	return nil
+}
+
+func (f *FileConfig) IsValid() (err error) {
+	// Path
+	if f.Path == "" {
+		return fmt.Errorf("invalid [Path] value: empty string")
+	}
+
+	// Permissions
```

Comment:
Missing validation here?

Reply:
Generally we call recursively into the members' `IsValid()`

Reply:
Oh, I saw cases where `IsValid()` wasn't being called recursively. Also, doesn't this mean that `IsValid()` will be called multiple times? Once for every `UnmarshalJSON()` of every parent object?

Reply:
Oh, this is kind of a mess. Not only do you call `IsValid()` during unmarshalling, but you also call it after the unmarshalling has completed. It should really be one or the other.

And if we do the validity checks during unmarshalling, then a parent object should not do validity checks on child objects with their own custom unmarshallers. Otherwise, there will be duplicate validity checking, which will unnecessarily slow down the code.

Reply:
Fixed.

Reply:
Yup, it's a bit ungainly, but perf isn't a huge issue for us since we generally only load these structs a single time. It allows us a few benefits. Obviously we run it while unmarshalling, but it also allows us to explicitly validate a structure at any time, and at any level. Also the unit tests are a bit more powerful since you can ensure there aren't any hidden issues further inside nested values, but also directly validate the IsValid() calls you care about.

I did the initial version of this stuff in a bit of a rush though, so happy for a better pattern.

Reply:
I think it is fine for now. The config files are pretty small. So, the fact that unmarhalling them isn't O(n) probably doesn't matter too much. If it ever becomes a problem, we can fix then.